scipy/optimize/tests/test_slsqp.py

```diff
@@ -135,6 +135,14 @@ def test_minimize_unbounded_combined(self):
                        jac=True, method='SLSQP', options=self.opts)
         assert_(res['success'], res['message'])
         assert_allclose(res.x, [2, 1])
+    
+    def test_minimize_unbounded_con_combined(self):
+        # Minimize, method='SLSQP': unbounded, combined constraint and Jacobian
+        # constraint: 2*x*y + 2*x - x**2 - 2*y**2 >= 0
+        con = {'type': 'ineq', 'fun': self.fun_and_jac, 'jac': True}
+        res = minimize(self.fun_and_jac, [-1.0, 1.0], args=(-1.0, ),
+                       jac=True, constraints=(con,), method='SLSQP', 
```

Comment:
Can you create an appropriate test with, and without, the `func_and_grad` functionality implemented in this PR, to show that the number of function evaluations decreases somewhere?

Reply:
I'm afraid that the `constr_nfev` and `constr_jfev` fields returned by 'trust-constr' that count the number of function/jacobian evaluations for each constraint won't change when using `jac=True` due to the logic of `MemoizeJac` (*). So I'm not really sure whether we should ignore this and if so, how could we test it properly?

To be precise, the constraint's jacobian function will still be called (and thus increases the `constr_jfev` counter) - the (only) difference is that it will use the jacobian computed in the previous evaluation of the constraint function. This can save a lot of computing time if the constraint function and the constraint jacobian share some expensive calculations.

The best I can think of would be something like this:

```python
class ConstraintFunctor:
    def __init__(self, fun, jac, con_and_jac=True):
        self.count_f = 0
        self.count_j = 0
        self.fun = fun
        self.jac = jac
        self.con_and_jac = con_and_jac
        
    def fun(self, x):
        self.count_f += 1
        if self.con_and_jac:
            return self.fun(x), self.jac(x)
        else:
            return self.fun(x)

    def jac(self, x):
        self.count_j += 1
        return self.jac(x)

cf1 = ConstraintFunctor(lambda x: x[0]**2, lambda x: 2*x[0], con_and_jac=True)
con1 = {'type': 'ineq', 'fun': cf1.fun, 'jac': cf1.jac}
res1 = minimize(x0=[-1.0, -1.0], constraints=(con1,), method="SLSQP")

cf2 = ConstraintFunctor(lambda x: x[0]**2, lambda x: 2*x[0], con_and_jac=False)
con2 = {'type': 'ineq', 'fun': cf2.fun, 'jac': True}
res2 = minimize(x0=[-1.0, -1.0], constraints=(con2,), method="SLSQP")
```

and then testing whether `cf1.count_f + cf1.count_j < cf2.count_f + cf2.count_j`. But this doesn't feel reasonable to me.

PS: (*) Afaik, the SLSQP algorithm doesn't count the function evaluations.